### PR TITLE
Rework api resources

### PR DESCRIPTION
**Pull request by gdraynz**
Changes:
* Splited all base resources into separate modules to clean things up (`nyuki/api/`)
* Resources can be put anywhere, but needs to be referenced from the nyuki that is using it.
* `version` is now `versions` and is a list of versions for this resource (always optional).
* The nyuki must now be accessed from the resources using `self.nyuki`.
```python
@resource('/my_resource', versions=['v1', 'v2'])
class MyApiResource:

    async def get(self, request):
        await self.nyuki.bus.publish({})


class MyNyuki(Nyuki):

    HTTP_RESOURCES = Nyuki.HTTP_RESOURCES + [MyApiResource]
```

### CODE REVIEW

File: nyuki/api/capabilities.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+from aiohttp.hdrs import METH_ALL
```
Comment (lhavel) [inline nyuki/api/capabilities.py]: The main module of a Python package is often named like the package itself: `api.py`

File: nyuki/api/capabilities.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    """
+    def decorated(cls):
+        cls.RESOURCE_CLASS = ResourceClass(cls, endpoint, versions, content_type)
+        return cls
+    return decorated
+
+
+class ResourceClass:
+
+    def __init__(self, cls, route, versions, content_type):
```
Comment (lhavel) [inline nyuki/api/capabilities.py]: Call it `route` or `endpoint` but don't use both. Technically this is more a `route` than an `endpoint`.

File: nyuki/api/capabilities.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        self._nyuki = nyuki
+        self._nyuki.register_schema(self.CONF_SCHEMA)
+        self._loop = self._nyuki.loop or asyncio.get_event_loop()
+        self._api = WebServer(self._loop)
+        self.host = None
+        self.port = None
+
+    @property
+    def capabilities(self):
+        return self.ENDPOINTS
```
Comment (lhavel) [inline nyuki/api/capabilities.py]: I guess this should be `self._nyuki.ENDPOINTS`

File: nyuki/api/webserver.py [hunk 4]
```diff
@@ -62,7 +40,7 @@ def _get_content_type(self, kwargs):
         return kwargs.get('content_type') or kwargs.get('headers', {}).get('content_type')
 
 
-class Api(object):
+class WebServer(object):
```
Comment (lhavel) [inline nyuki/api/webserver.py]: Already implicitely inherits from `object` (just like the other classes you wrote).

File: nyuki/api/webserver.py [hunk 5]
```diff
@@ -121,10 +99,9 @@ def router(self):
     Transform the request data to be passed through a capability and
     convert the result into a web response.
     """
-
     async def middleware(request):
         # Ensure a content-type check is necessary
-        if hasattr(capa_handler, 'CONTENT_TYPE') and await request.text():
+        if getattr(capa_handler, 'CONTENT_TYPE', None) and await request.text():
```
Comment (lhavel) [inline nyuki/api/webserver.py]: Why did you change `hasattr` to `getattr`?

File: nyuki/nyuki.py [hunk 6]
```diff
@@ -1,28 +1,27 @@
...
+from .logs import DEFAULT_LOGGING
+from .services import ServiceManager
+from .websocket import WebHandler
 
 
 log = logging.getLogger(__name__)
 
 
-class Nyuki(metaclass=CapabilityHandler):
+class Nyuki:
```
Comment (lhavel) [inline nyuki/nyuki.py]: I like that metaclass removed! +1

File: nyuki/nyuki.py [hunk 7]
```diff
@@ -41,6 +40,14 @@ class Nyuki(metaclass=CapabilityHandler):
         "type": "object",
         "required": ["log"]
     }
+    # API endpoints
+    ENDPOINTS = [
```
Comment (lhavel) [inline nyuki/nyuki.py]: See my comment about the naming rules.

Comment (lhavel) [inline nyuki/nyuki.py]: Do you really need to redefine the whole list when you create a new nyuki? If yes, we should find a better way to do it.

File: nyuki/api/webserver.py (hunk 5 above)
Comment (gdraynz) [inline nyuki/api/webserver.py]: `content-type` can be set to `None` to avoid any content-type verification upon receiving a request (useful when we don't care about any data we receive)

File: nyuki/nyuki.py (hunk 7 above)
Comment (gdraynz) [inline nyuki/nyuki.py]: In `WorkflowNyuki` we use this method to append resources from the base nyuki and others:
```python
HTTP_RESOURCES = Nyuki.HTTP_RESOURCES + [MyOtherResource, ]
```

File: nyuki/api/capabilities.py (hunk 2 above)
Comment (gdraynz) [inline nyuki/api/capabilities.py]: Done, there was indeed a confusion around naming

**Comment by lerignoux** [comment]
better to use extends than +

**Comment by lerignoux** [comment]
if you want workflow capabilities HTTP_RESOURCES = WorkflowNyuki.HTTP_RESOURCES must be used !!
